### PR TITLE
refactor: slightly improve `NodeCommunication`s

### PR DESCRIPTION
**Pull request by mlegner**
- Avoid creating `NodeCommunication`s for all nodes when only a subset is required.
- Only return an error if there is a good reason and the case is actually possible.

### CODE REVIEW

File: crates/walrus-service/src/client/communication.rs [hunk 1]
```diff
@@ -78,19 +77,20 @@ impl<'a> NodeCommunication<'a> {
...
             %node_index,
             %config.max_node_connections,
             "initializing communication with node"
         );
-        ensure!(
-            !node.shard_ids.is_empty(),
-            ClientErrorKind::InvalidConfig.into()
-        );
-        Ok(Self {
+        Some(Self {
```
Comment (giac-mysten) [inline crates/walrus-service/src/client/communication.rs:93]: Important question: IIRC we had decided to go with errors because, for a valid config, there should never be a node without shards. Now this is relaxed. Do you foresee cases in which this is useful?

Comment (mlegner) [reply crates/walrus-service/src/client/communication.rs:93]: Hmm... I don't have a strong opinion on this. In practice, there shouldn't be a difference in behavior anyway as we check for empty nodes when constructing the `Committee`: https://github.com/MystenLabs/walrus/blob/ml/refactor-node-comms/crates/walrus-sui/src/types.rs#L287

For me then, the errors just make it a bit hard to follow when errors are returned and which (because of the opaque `NodeError` type). Given invariants on `Committee` I would also be open to just panic in some cases.

Comment (mlegner) [reply crates/walrus-service/src/client/communication.rs:93]: I explicitly added a comment about the invariants on `Committee` in e9b945ba96cfaba5d9b40359d4506390558b94b1.

Comment (giac-mysten) [reply crates/walrus-service/src/client/communication.rs:93]: works for me!
